gqlalchemy/transformations.py

```diff
@@ -0,0 +1,152 @@
+# limitations under the License.
+
+import logging
+import multiprocessing as mp
+from typing import Any, Dict, Iterator, List, Union
+
+import mgclient
+import networkx as nx
+
+from gqlalchemy import Memgraph
+from gqlalchemy.utilities import to_cypher_labels, to_cypher_properties
+
+__all__ = ("nx_to_cypher", "nx_graph_to_memgraph_parallel")
+
+
+class NetworkXGraphConstants:
+    LABELS = "labels"
+    TYPE = "type"
+    ID = "id"
+
+
+def nx_to_cypher(graph: nx.Graph) -> Iterator[str]:
+    """Generates a Cypher queries for creating graph"""
+
+    yield from _nx_nodes_to_cypher(graph)
+    yield from _nx_edges_to_cypher(graph)
+
+
+def nx_graph_to_memgraph_parallel(
+    graph: nx.Graph,
+    host: str = "127.0.0.1",
+    port: int = 7687,
+    username: str = "",
+    password: str = "",
+    encrypted: bool = False,
+) -> None:
+    """Generates a Cypher queries and inserts data into Memgraph in parallel"""
+    num_of_processes = mp.cpu_count() // 2
+    _check_for_index_hint(
+        host,
+        port,
+        username,
+        password,
+        encrypted,
+    )
+    for queries_gen in [_nx_nodes_to_cypher(graph), _nx_edges_to_cypher(graph)]:
+        queries = list(queries_gen)
+        chunk_size = len(queries) // num_of_processes
+        processes = []
+        for i in range(num_of_processes):
+            process_queries = queries[i * chunk_size : chunk_size * (i + 1)]
+            processes.append(
+                mp.Process(
+                    target=_insert_queries,
+                    args=(
+                        process_queries,
+                        host,
+                        port,
+                        username,
+                        password,
+                        encrypted,
+                    ),
+                )
+            )
+        for p in processes:
+            p.start()
+        for p in processes:
+            p.join()
+
+
+def _check_for_index_hint(
+    host: str = "127.0.0.1",
+    port: int = 7687,
+    username: str = "",
+    password: str = "",
+    encrypted: bool = False,
+):
+    memgraph = Memgraph(host, port, username, password, encrypted)
+    indexes = memgraph.get_indexes()
+    if len(indexes) == 0:
+        logging.getLogger(__file__).warning(
+            f"Be careful you do not have any indexes set up, the queries will take longer than expected!"
+        )
+
+
+def _insert_queries(queries: List[str], host, port, username, password, encrypted) -> None:
+    """Used by multiprocess insertion of nx into memgraph, works on a chunk of queries."""
+    memgraph = Memgraph(host, port, username, password, encrypted)
+    while len(queries) > 0:
+        try:
+            query = queries.pop()
+            memgraph.execute_query(query)
+        except IndexError:
+            break
+        except mgclient.DatabaseError as e:
+            queries.append(query)
+            logging.getLogger(__file__).warning(f"Ignoring database error: {str(e)}")
```

Comment:
call to `str` in an f-string? Does it make a difference?

Reply:
Nope it does not will remove